### PR TITLE
Serde-rs deserializer support

### PR DESCRIPTION
**Pull request by Relrin**

### CODE REVIEW

**Comment by dtolnay** [comment]
[Serde 0.8.7](https://github.com/serde-rs/serde/releases/tag/v0.8.7) provides its own `forward_to_deserialize!` macro so you don't need to copy forward.rs. [Here is example usage from serde_json](https://github.com/serde-rs/json/pull/148/files).

**Comment by Relrin** [comment]
Thanks for this proposal. It looks very helpful for me 👍

File: bert/src/deserializers.rs [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+        match self.length {
+            Some(len) => (len, self.length),
+            None => (0, Some(0))
+        }
+    }
+}
+
+
+struct BertBigNumberVisitor<'a, R: 'a + Read> {
+    de: &'a mut Deserializer<R>,
```
Comment (oli-obk) [inline bert/src/deserializers.rs]: remove these fields. This Visitor should not have state.

File: bert/src/deserializers.rs [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+        BertBigNumberVisitor { de: de, header: header }
+    }
+}
+
+
+impl<'a, R: 'a + Read> de::Visitor for BertBigNumberVisitor<'a, R> {
+    type Value = BertBigInteger;
+
+    #[inline]
+    fn visit_newtype_struct<D>(&mut self, _: &mut D) -> StdResult<BertBigInteger, D::Error>
```
Comment (oli-obk) [inline bert/src/deserializers.rs]: don't ignore the deserializer, use it, don't pass the deserializer through a field.

File: bert/src/deserializers.rs [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+        Ok(BertBigInteger(bigint))
+    }
+}
+
+
+impl Deserialize for BertBigInteger {
+    fn deserialize<D>(deserializer: &mut D) -> StdResult<BertBigInteger, D::Error>
+        where D: de::Deserializer
+    {
+        deserializer.deserialize_newtype_struct(
```
Comment (oli-obk) [inline bert/src/deserializers.rs]: just call `deserializer.deserialize(BertBigNumberVisitor)`

File: bert/src/deserializers.rs [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+
+
+impl<'a, R: 'a + Read> de::Visitor for BertBigNumberVisitor<'a, R> {
+    type Value = BertBigInteger;
+
+    #[inline]
+    fn visit_newtype_struct<D>(&mut self, _: &mut D) -> StdResult<BertBigInteger, D::Error>
+        where D: de::Deserializer
+    {
+        let n = match self.header {
```
Comment (oli-obk) [inline bert/src/deserializers.rs]: You should not be accessing the deserializer directly. Instead, use `Deserialize::deserialize` calls to obtain the correct value, in this case, `let v: Vec<u8> = Deserialize::deserialize(deserializer)`

File: bert/src/deserializers.rs [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+    ) -> Result<V::Value> {
+        let length = self.read_i32::<BigEndian>().unwrap() as usize;
+        visitor.visit_seq(BinarySeqVisitor::new(self, Some(length)))
+    }
+
+    #[inline]
+    fn parse_big_integer<V: Visitor>(
+        &mut self, header: u8, mut visitor: V
+    ) -> Result<V::Value> {
+        visitor.visit_newtype_struct(BertBigNumberVisitor::new(self, header))
```
Comment (oli-obk) [inline bert/src/deserializers.rs]: Here you need to pass `BertBigNumberDeserializer::new(self, header)`.

File: bert/src/deserializers.rs [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+            111 => try!(self.de.read_u32::<BigEndian>())
+        };
+        let sign_int = try!(self.de.read_u8());
+        let sign = match sign_int {
+            0 => Sign::Plus,
+            _ => Sign::Minus
+        };
+        let bytes = try!(self.de.read_exact(n as u64));
+        let bigint = BigInt::from_bytes_le(sign, bytes.as_ref());
+        Ok(BertBigInteger(bigint))
```
Comment (oli-obk) [inline bert/src/deserializers.rs]: just do `let v: Vec<u8> = try!(Deserializer::deserialize(deserializer)); and then do all the above code code inside`BertBigNumberDeserializer`.

**Review by oli-obk — CHANGES_REQUESTED** [review]
I was wrong. You don't need a second `Visitor`, but a `Deserializer` whose sole purpose is to deserialize BigNum

File: bert/src/deserializers.rs (hunk 5 above)
Comment (Relrin) [reply bert/src/deserializers.rs]: I don't quite understand what is the `BertBigNumberDeserializer::new(self, header)`. Maybe you mean BertBigNumberVisitor there? As far as I remember, for each custom type we must to implement `Deserialize` and `Visitor` traits for the concrete type which we want to deserialize in the result, isn't it? Perhaps you me mean to implement similar deserializer like default and use it?

Comment (Relrin) [reply bert/src/deserializers.rs]: And yet another question. This suggestion too much different from a suggestions which described in the Serde docs (like [this](https://serde.rs/deserialize-map.html)).
@dtolnay Why this approaches too different from the documentation? In my case only what I need is just read few bytes (in the worst case few bytes additionally) and put all this values into `num::bigint::BigInt` structure. Occasionally, it makes me thing that we doing too much actions to make a simple task.
In compare with Serde serializers functionality deserializers look too complex and harder in maintenance in further. That big differences (sometimes too much) just blows my mind :(

**Pull request closed without merging**